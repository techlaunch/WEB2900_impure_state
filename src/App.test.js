import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import App from './App';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
});

it('immutably adds items to state', () => {
  Math.random = (() => {
    let count = 0;
    return () => count++;
  })();

  const cmp = mount(<App/>);
  expect(toJson(cmp)).toMatchSnapshot();
  cmp.find('#immutable-add').simulate('click');
  expect(toJson(cmp)).toMatchSnapshot();
  cmp.find('#immutable-add').simulate('click');
  expect(toJson(cmp)).toMatchSnapshot();

  expect(cmp.find('li').length).toBe(2);
});

it('mutably adds items to state but does not re-render', () => {
  const cmp = mount(<App/>);
  expect(toJson(cmp)).toMatchSnapshot();
  cmp.find('#mutable-add').simulate('click');
  expect(toJson(cmp)).toMatchSnapshot();
  cmp.find('#mutable-add').simulate('click');
  expect(toJson(cmp)).toMatchSnapshot();

  expect(cmp.find('li').length).toBe(0);
});

it('renders after an immutable add', () => {
  const cmp = mount(<App/>);
  expect(toJson(cmp)).toMatchSnapshot();
  cmp.find('#mutable-add').simulate('click');
  expect(cmp.find('li').length).toBe(0);
  cmp.find('#immutable-add').simulate('click');
  expect(cmp.find('li').length).toBe(2);
});
