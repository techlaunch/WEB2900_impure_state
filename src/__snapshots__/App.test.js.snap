// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`immutably adds items to state 1`] = `
<App>
  <div>
    <button
      id="immutable-add"
      onClick={[Function]}
    >
      Add item immutably (good)
    </button>
    <button
      id="mutable-add"
      onClick={[Function]}
    >
      Add item mutably (bad)
    </button>
    <ItemList
      items={Array []}
    >
      <ul />
    </ItemList>
  </div>
</App>
`;

exports[`immutably adds items to state 2`] = `
<App>
  <div>
    <button
      id="immutable-add"
      onClick={[Function]}
    >
      Add item immutably (good)
    </button>
    <button
      id="mutable-add"
      onClick={[Function]}
    >
      Add item mutably (bad)
    </button>
    <ItemList
      items={
        Array [
          Object {
            "id": 0,
            "value": 0,
          },
        ]
      }
    >
      <ul>
        <li>
          0
        </li>
      </ul>
    </ItemList>
  </div>
</App>
`;

exports[`immutably adds items to state 3`] = `
<App>
  <div>
    <button
      id="immutable-add"
      onClick={[Function]}
    >
      Add item immutably (good)
    </button>
    <button
      id="mutable-add"
      onClick={[Function]}
    >
      Add item mutably (bad)
    </button>
    <ItemList
      items={
        Array [
          Object {
            "id": 0,
            "value": 0,
          },
          Object {
            "id": 1,
            "value": 1,
          },
        ]
      }
    >
      <ul>
        <li>
          0
        </li>
        <li>
          1
        </li>
      </ul>
    </ItemList>
  </div>
</App>
`;

exports[`mutably adds items to state but does not re-render 1`] = `
<App>
  <div>
    <button
      id="immutable-add"
      onClick={[Function]}
    >
      Add item immutably (good)
    </button>
    <button
      id="mutable-add"
      onClick={[Function]}
    >
      Add item mutably (bad)
    </button>
    <ItemList
      items={Array []}
    >
      <ul />
    </ItemList>
  </div>
</App>
`;

exports[`mutably adds items to state but does not re-render 2`] = `
<App>
  <div>
    <button
      id="immutable-add"
      onClick={[Function]}
    >
      Add item immutably (good)
    </button>
    <button
      id="mutable-add"
      onClick={[Function]}
    >
      Add item mutably (bad)
    </button>
    <ItemList
      items={
        Array [
          Object {
            "id": 0,
            "value": 2,
          },
        ]
      }
    >
      <ul />
    </ItemList>
  </div>
</App>
`;

exports[`mutably adds items to state but does not re-render 3`] = `
<App>
  <div>
    <button
      id="immutable-add"
      onClick={[Function]}
    >
      Add item immutably (good)
    </button>
    <button
      id="mutable-add"
      onClick={[Function]}
    >
      Add item mutably (bad)
    </button>
    <ItemList
      items={
        Array [
          Object {
            "id": 0,
            "value": 2,
          },
          Object {
            "id": 1,
            "value": 3,
          },
        ]
      }
    >
      <ul />
    </ItemList>
  </div>
</App>
`;

exports[`renders after an immutable add 1`] = `
<App>
  <div>
    <button
      id="immutable-add"
      onClick={[Function]}
    >
      Add item immutably (good)
    </button>
    <button
      id="mutable-add"
      onClick={[Function]}
    >
      Add item mutably (bad)
    </button>
    <ItemList
      items={Array []}
    >
      <ul />
    </ItemList>
  </div>
</App>
`;
